hypervisor/common/ptdev.c

```diff
@@ -116,6 +129,29 @@ static void ptdev_interrupt_handler(__unused uint32_t irq, void *data)
 	struct ptdev_remapping_info *entry =
 		(struct ptdev_remapping_info *) data;
 
+	/*
+	 * "interrupt storm" detection & delay intr injection just for UOS
+	 * pass-thru devices, collect its data and delay injection if needed
+	 */
+	if (!is_vm0(entry->vm)) {
+		entry->intr_count++;
+
+		/* if delta > 0, it needs to delay the irq */
+		if (entry->vm->intr_inject_delay_delta > 0UL) {
+
+			/*
+			 * if delay timer fire_tsc == 0, add it into timer
+			 * list, else it was added already, just ignore
+			 */
+			if (entry->intr_delay_timer.fire_tsc == 0UL) {
+				entry->intr_delay_timer.fire_tsc = rdtsc() +
```

Comment:
here,  it is used to avoid the timer is added reproduced. if same interrupt entry's timer is added once, just when it is handled, it can be added into timer list again.

Reply:
IMHO, the "if" condition is OK. 
And 
if (!list_empty(&entry->intr_delay_timer.node))
 would be better

Reply:
Even if we do so, there are Windows that may violates the assumption. 
If we assume the hardware itself is trusted, it won't generate 2nd edge triggered interrupt.

Let us state it as decision logic.

Reply:
as Eddie's note, I'm thinking we could not operate timer list in ISR, it could happen when timer is operated by the same CPU. (now timer is per CPU).  it could be a risk. I'll check more.

Reply:
Good point!  I think we should move back to original  TSC solution. @anthonyzxu

Reply:
you can use softirq to add timer.

Reply:
It could. But why we need this kind of complexity? TSC just works fine and is much simpler.  In the case of interrupt storm, we can simply kill the guest or hang the guest.